### DISCUSSION TITLE
udev rule/script to add Bluealsa device 'on the fly'

### DISCUSSION THREAD

**Discussion by gearhead** [Arkq/bluez-alsa#527] [Q&A]
I mentioned this a while ago and was asked to post when I figured out something. Well, I figured out how to tell it when I connect/disconnect with udev rules. I noticed that all audio devices I have added all advertise as 'hci_uart_bcm" and "bluetooth". These rules do work and call my script correctly. Now to flesh out the script to parse out the type of device with bluealsa-aplay as the second line returns if it is a source or a sink. Is there a better way of doing this? Any help appreciated. 
```
SUBSYSTEM=="bluetooth", DRIVERS=="hci_uart_bcm", ACTION=="add", KERNEL=="hci[0-9]*", RUN+="/srv/http/command/bt-connect.sh start"
SUBSYSTEM=="bluetooth", DRIVERS=="hci_uart_bcm", ACTION=="remove", KERNEL=="hci[0-9]*", RUN+="/srv/http/command/bt-connect.sh stop"
```
this calls my script and I am terrible at writing and debugging scripts, but this is where I am now. It does not actually yet test for source or sink, yet, but I am working on it. 
```
#!/usr/bin/bash
# bt-connect.sh
# shell script to switch off mpd so that a source connected
# will play through alsa system to the default output.
# also if a speaker/headset is added, it will set it as the default
# output in mpd.conf
case $1 in
"start")
    type=$(bluealsa-aplay -L | grep trusted | cut -d ',' -f 3 | tr -d -d " \t\n\r")
       case $type in
       "playback") # attached speaker
       mpc stop
       # set default output in mpd.conf to BT
       ;;
       "capture") # attached a source
       mpc stop
       # this is enough to allow a device to play through alsa
       ;;
      esac
"stop")
   # remove the BT default if it was added to mpd.conf
   mpc play ;;
esac
```

I can run it from cli 'bt-connect.sh start' and it does what is is supposed to but when triggered from the udev rule on start, it does not. If I disconnect, the second case (stop) works. If simplify the rule to only be start and stop and remove the nested 'type' case switch, it works from udev... 
Any ideas?

**Comment by gearhead** [comment]
So I read a lot about udev rules and the need for the full path in everything. I changed the one assignment to be:
`type=$(/usr/bin/bluealsa-aplay -L | /usr/bin/grep trusted | /usr/bin/cut -d ',' -f 3 | /usr/bin/tr -d -d " \t\n\r")`
it still runs fine from the cli, but as a udev rule it does not assign anything to 'type'. 

Any ideas why it does not run? all the functions I am running are in /usr/bin (bluealsa-aplay, grep, cut, tr) my guess is that bluealsa-aplay will not run from a udev rule. I tried sending its output to a log and it always reports nothing. Why does this not run from the udev rule?

**Comment by borine** [comment]
I see a number of potential issues with your udev rules and script:
```
DRIVERS=="hci_uart_bcm"
```
This rule will work only with Broadcom UART adapters (eg RPi on-board adapter), but not with USB adapters, nor with adapters by any other manufacturer. It would be easier to move to other machines if made more generic. (see my suggestion below).

But the main issue is with the script. Your script is invoked immediately any bluetooth device connects, which means it most likely runs *before* BlueALSA has created any PCMs. So bluealsa-aplay -L reports nothing - exactly the result you are seeing.

You need to *wait* for the PCMs to be created before listing them. I recommend using `bluealsa-cli monitor` for this purpose. Here is a suggested alternative - there will be many other ways of achieving this, but here is my shot at it:

/etc/udev/rules.d/90-bt-connect.rules:
```
SUBSYSTEM=="bluetooth", ENV{DEVTYPE}=="link", ACTION=="add", RUN+="/srv/http/command/bt-connect.sh add"
SUBSYSTEM=="bluetooth", ENV{DEVTYPE}=="link", ACTION=="remove", RUN+="/srv/http/command/bt-connect.sh remove"
```

/srv/http/command/bt-connect.sh:
```shell
#!/bin/sh
# bt-connect.sh
# shell script to switch off mpd so that a source connected
# will play through alsa system to the default output.
# also if a speaker/headset is added, it will set it as the default
# output in mpd.conf
case "$1" in
add)
	# wait for PCM(s) to be created, with absolute timeout of 5 seconds.
	timeout 5 bluealsa-cli -q monitor | grep -q -m 1 ^PCMAdded || exit 0
	# 
	case "$(bluealsa-cli -q list-pcms | grep -o -E '(sink|source)$')" in
	*sink*) # attached speaker(s)
		mpc stop
		# set default output in mpd.conf to BT
		;;
	*source*) # attached phone
		mpc stop
		# this is enough to allow a device to play through alsa
		;;
	esac
	;;
remove)
	# remove the BT default if it was added to mpd.conf
	mpc play
	;;
esac
```

**Reply by gearhead** [reply]
@borine Thanks for the help. I do not have bluealsa-cli... I'll need to find out where it is and build it as well. I build and use bluez-alsa 3.1.0. I'll keep you posted.

**Reply by borine** [reply]
configure bluealsa with `--enable-cli` - the manual page is here https://github.com/Arkq/bluez-alsa/blob/master/doc/bluealsa-cli.1.rst

**Comment by gearhead** [comment]
I rebuilt with `--enable-cli` and was able to get the `bluealsa-cli` executable. 
```
# which bluealsa-cli
/usr/bin/bluealsa-cli
```
Changed to your suggested rules and script and it does a very similar thing. I added a few log writes and see that the rule is working for the phone I am playing with. I send 'add' and get that far, but nothing past the 'timeout' line runs when I connect. If I manually 'mpc stop' then 'remove' by disconnecting from the phone the 'remove' part of the script runs fine and turns mpd back on. 
In case it is important, this is running on a RPi3B+ right now. It appears that timeout will not run this way. I did add /usr/bin to all of the calls and verified that is where the executables are just in case udev needs the full path (I have read that can be a problem).

my edited line that is not executed:
`/usr/bin/timeout 5 /usr/bin/bluealsa-cli -q monitor | /usr/bin/grep -q -m 1 ^PCMAdded || exit 0`

any further ideas appreciated.

**Reply by gearhead** [reply]
There is something that does not work with the 'timeout' line. When I run bluealsa-cli, It looks like this:
```
#  bluealsa-cli -q monitor
PCMRemoved /org/bluealsa/hci0/dev_XX_XX_XX_XX_XX_XX/a2dpsnk/source
PCMAdded /org/bluealsa/hci0/dev_XX_XX_XX_XX_XX_XX/a2dpsnk/source
PCMRemoved /org/bluealsa/hci0/dev_XX_XX_XX_XX_XX_XX/a2dpsnk/source
```
I am not a cli expert, so am poking in the dark a bit, but I get no response when I add and remove a device with this:
`# bluealsa-cli -q monitor | grep -q -m 1 ^PCMAdded`

When I run the whole line, it does exit my ssh session when I add a phone, so the line seems OK. but something doe snot yet work... Still plugging. 
`/usr/bin/timeout 5 /usr/bin/bluealsa-cli -q monitor | /usr/bin/grep -q -m 1 ^PCMAdded || exit 0`

I'll see if I can figure out what is going on . I appreciate your help!

**Reply by borine** [reply]
Hmm, I can think of 2 possibilities:
- a RPi3B+ needs more than 5 seconds to generate the PCMs - try increasing the timeout, say 10 secs.
- udev is setting its own time limit for actions, and is killing your script early. This is quite possible, as it is not intended to be used for starting long-running services. In which case, you will have to search more generally for udev advice

I personally don't like using udev this way. BlueALSA PCMs are not kernel objects so using a kernel event monitoring system to manage them just seems wrong to me. I prefer to run my own monitoring daemon that listens for D-Bus events which is a more natural place to manage BlueALSA objects.

**Reply by gearhead** [reply]
My guess is it is udev that I am fighting. I tried 10 sec and it is the same result. It never gets past that 'timeout' line.

I agree with you that udev may not be the 'best way' to do this.  I have touched on dbus, but it seems pretty opaque to me, especially on cli. Can you share the monitoring daemon that you use? We are trying to add bluetooth to Runeaudio and as you can see I am a bit over my head on this. It just seems like BT should be part of Runeaudio and we are struggling to try to get it in there.

As you probably already know, we need to add the bluealsa output to mpd and restart it to be able to use it with mpd. With a source, adding a phone, we just need to stop mpd so alsa is clear to use the default output. We have not considered using both BT input and output but wonder if an RPi can actually process both streams at the same time or of bluealsa can actually do that.

**Reply by borine** [reply]
I have been using a scripted solution with some success, shared here: https://github.com/borine/bluealsa-monitor
I am now working on a new solution written in C, reusing code from bluez-alsa. Its not ready for release yet, but my plan is to offer it upstream to @Arkq for inclusion in the bluez-alsa project eventually. If it is considered out-of-scope for bluez-alsa then I will probably share it in a new project. As a work-in-progress you can see it as a bluez-alsa utility here:https://github.com/borine/bluez-alsa/tree/autoconfig

**Reply by borine** [reply]
> I get no response when I add and remove a device with this:
```# bluealsa-cli -q monitor | grep -q -m 1 ^PCMAdded```

Right, my bad, you need the latest bluez-alsa code for that. In release 3.1.0 bluealsa-cli uses fully buffered output to pipes. So you need to fixup stdout to line-buffered mode  before running bluealsa-cli. Try:
```
stdbuf -oL bluealsa-cli -q monitor | grep -q -m 1 ^PCMAdded
```
There will be no output (`grep -q`), the command will terminate as soon a a BlueALSA PCM is added.

[EDIT] this is a third possible explanation of the udev script failure - try again but with the line:
```
timeout 5 stdbuf -oL bluealsa-cli -q monitor | grep -q -m 1 ^PCMAdded || exit 0
```

**Reply by gearhead** [reply]
The first option worked:
`timeout 5 stdbuf -oL bluealsa-cli -q monitor | grep -q -m 1 ^PCMAdded || exit 0`
Thanks for the help! I will look at your companion and will try to add it instead of udev, but this now works as I was hoping. It is a start.

**Reply by gearhead** [reply]
@borine So, now that I have the bluealsa-monitor running and a reasonably current bluealsa, I have returned to my udev script to set things up. I know you mentioned that this was a bit crude before. Could you point me in a better direction? 
As it is, I have scripts that I can run manually to add the bluealsa device for mpd and also remove it. I put this code in the script that is triggered by udev. I have it removing the bluealsa device from mpd when it is removed, but I cannot get it to 'automatically' add the device to the mpd config. I have a hunch it is this line that is causing the problem, but am unsure how to fix it...
`/usr/bin/bluealsa-cli -q list-pcms | grep -o -E '(sink|source)$'`
When I run this line from cli after my headset is added, I get this:
```
# /usr/bin/bluealsa-cli -q list-pcms | grep -o -E '(sink|source)$'
source
sink
sink
```
so it looks like it parses that it added a source and skips the sinks.

The full output is:
```
# /usr/bin/bluealsa-cli -q list-pcms
/org/bluealsa/hci0/dev_70_BF_92_D8_82_84/a2dpsrc/sink
/org/bluealsa/hci0/dev_70_BF_92_D8_82_84/hfpag/source
/org/bluealsa/hci0/dev_70_BF_92_D8_82_84/hfpag/sink
```
I added another loop through grep to get just one.
```
# /usr/bin/bluealsa-cli -q list-pcms | grep a2dp
/org/bluealsa/hci0/dev_70_BF_92_D8_82_84/a2dpsrc/sink
# /usr/bin/bluealsa-cli -q list-pcms | grep a2dp | grep -o -E '(sink|source)$'
sink
```
This is the script that is triggered when a bt device is added/removed. Remove works. If the bluealsa is manually added to mpd.conf. it is removed by this script automatically when I disconnect the headphones.  When I connect, though, nothing happens and the journal acts like it is not run at all. It should turn off mpd and that never happens. 
```
#!/usr/bin/sh
# bt-connect.sh
# shell script to switch off mpd so that a source connected
# will play through alsa system to the default output.
# also if a speaker/headset is added, it will set it as the default
# output in mpd.conf
# update disconnect works, connect does not.
file="/etc/mpd.conf"
case "$1" in
add)
# wait for PCM(s) to be created, with absolute timeout of 5 seconds.
timeout 5 stdbuf -oL bluealsa-cli -q monitor | grep -q -m 1 ^PCMAdded || exit 0
    case "$(/usr/bin/bluealsa-cli -q list-pcms | grep a2dp | grep -o -E '(sink|source)$')" in
    sink) # attached speaker(s)
        mpc stop
        systemctl stop mpd
        sed -i '/audio_output/,/}/ s/enabled/#&/ ' $file
        mapfile -t bts < <( aplay -L )
        lines=$(aplay -L | wc -l )
        let lines='lines-1'
        if [ $lines -ne 0 ]; then
            for i in $( eval echo {0..$lines..2});do
                if echo "${bts[$i]}" | grep -q a2dp ; then
                    name=$(echo ${bts[$i+1]} | cut -d "," -f 1)
                    device=$(echo ${bts[$i]})
                    echo "audio_output {" >> $file
                    echo -ne "        name            \"" >> $file
                    echo -ne $name >> $file
                    echo "\"" >> $file
                    echo "        type            \"alsa\"" >> $file
                    echo -ne "        device          \"" >> $file
                    echo -ne $device >> $file
                    echo "\"" >> $file
                    echo "        mixer_type      \"software\"" >> $file
                    echo "        auto_resample   \"no\"" >> $file
                    echo "        auto_format     \"no\"" >> $file
                    echo "        enabled         \"yes\"" >> $file
                    echo "}" >> $file
                fi
            done
        fi
        systemctl start mpd.socket
        # set default output in mpd.conf to BT
        ;;
    source) # attached phone
        mpc stop
        # this is enough to allow a device to play through alsa
        ;;
    esac
    ;;
remove)
    mpc stop
    blue=$(grep -n bluealsa $file | cut -d ":" -f 1)
    # line where we found "bluealsa" if found
    if [ $blue -ne 0 ]
    then
        systemctl stop mpd
        mapfile -t ao < <( grep -n audio_output $file | cut -d ":" -f 1 )
        numouts=${#ao[@]}
        mapfile -t bracket < <( grep -n } $file | cut -d ":" -f 1 )
        numbrkts=${#bracket[@]}
        let numbrkts='numbrkts-1'
        let numouts='numouts-1'
        # find which bracket line number is greater than bluealsa and which ao is less than
        let i=0
        while [ $blue -gt ${bracket[$i]} ]; do
              (( i++))
        done
        # i is the index of the bracket past bluealsa
        end=${bracket[$i]}
        # now lets find the ao which is just greater than bluealsa
        let k=0
        # if blue is gt this ao and less then brkt, we found it. 
        for j in $(eval echo {0..$numouts});do
             if test $blue -gt ${ao[$j]}; then
             (( k++ ))
             fi
        done
        begin=${ao[$k-1]}
        # remove the BT default if it was added to mpd.conf 
        sed -i "$begin,$end d" $file
        systemctl start mpd.socket
   fi
    mpc play
    ;;
esac
```
This is the journal showing me disconnecting then reconnecting:
```
Feb 13 22:29:01 rune64 bluealsa[405]: /usr/bin/bluealsa: E: BT socket disconnected: Transport endpoint is not connected
Feb 13 22:29:01 rune64 bluealsa[405]: /usr/bin/bluealsa: E: Couldn't release transport: GDBus.Error:org.freedesktop.DBus.Error.UnknownObject: Method "Release" with signature "" on interface "org.bluez.MediaTransport1" doesn't exist
Feb 13 22:29:01 rune64 bluealsa[405]: /usr/bin/bluealsa: E: Couldn't release transport: GDBus.Error:org.freedesktop.DBus.Error.UnknownObject: Method "Release" with signature "" on interface "org.bluez.MediaTransport1" doesn't exist
Feb 13 22:29:03 rune64 systemd[1]: Stopping Music Player Daemon...
Feb 13 22:29:04 rune64 systemd[1]: mpd.service: Deactivated successfully.
Feb 13 22:29:04 rune64 systemd[1]: Stopped Music Player Daemon.
Feb 13 22:29:04 rune64 systemd[1]: mpd.service: Consumed 25.735s CPU time.
Feb 13 22:29:04 rune64 systemd[1]: mpd.socket: Deactivated successfully.
Feb 13 22:29:04 rune64 systemd[1]: Closed MPD Socket.
Feb 13 22:29:04 rune64 systemd[1]: Listening on MPD Socket.
Feb 13 22:29:04 rune64 systemd[1]: Starting Music Player Daemon...
Feb 13 22:29:05 rune64 systemd[1]: Started Music Player Daemon.
Feb 13 22:29:21 rune64 bluetoothd[302]: src/profile.c:ext_connect() Headset Voice gateway failed connect to 70:BF:92:D8:82:84: Connection refused (111)
Feb 13 22:29:21 rune64 bluealsa[405]: /usr/bin/bluealsa: W: Unsupported AT message: SET: command:+CLIP, value:1
Feb 13 22:29:21 rune64 kernel: input: Jabra Evolve2 65 (AVRCP) as /devices/virtual/input/input2
Feb 13 22:29:21 rune64 systemd-logind[1048]: Watching system buttons on /dev/input/event1 (Jabra Evolve2 65 (AVRCP))
```
it is connected. 
```
# /usr/bin/bluealsa-cli -q list-pcms | grep a2dp | grep -o -E '(sink|source)$'
sink
```
I'd appreciate any feedback on this.

**Comment by borine** [comment]
For the issue with SCO PCMs breaking your script, either run `bluealsa` without the SCO profiles if you don't use them, or modify the `timeout` pipeline to filter only A2DP. In the latter case you could get the PCM path from the same command. Something like (not tested):
```
event=$(timeout 5 stdbuf -oL bluealsa-cli -q monitor | grep -m 1 '^PCMAdded.*a2dp' || exit 0)
case "$event" in
	*sink)
		;;
	*source)
		;;
esac
```
 I suggested `bluealsa-monitor` as an example of an *alternative* to udev events for detecting addition and removal of BlueALSA PCMs. I can't help any further with udev as it really is not of interest to me.
